### ISSUE TITLE
TypeVar bound to Protocol with forward reference causes false positive assignment errors since v1.1.385

### ISSUE THREAD

**Issue by ibeckermayer** [microsoft/pyright#10794]
**Describe the bug**
Starting in pyright v1.1.385, functions that use a TypeVar bound to a Protocol are incorrectly reporting assignment and return type errors when the protocol method returns a forward reference to the protocol itself. This worked correctly in v1.1.384 and earlier versions.

The error message shows `Type "ProtocolName" is not assignable to declared type "T@function_name"` even when the types should be compatible.

**Code or Screenshots**
Minimal reproduction case:

```python
from typing import Protocol, TypeVar

# Two different classes that both have the same method
class TypeA:
    def process(self, arg: str) -> "TypeA":
        return self

class TypeB:
    def process(self, arg: str) -> "TypeB":
        return self

# Protocol that both classes should satisfy
class Processor(Protocol):
    def process(self, arg: str) -> "Processor": ...

# TypeVar bound to the protocol
T = TypeVar("T", bound=Processor)

# Function that should work but triggers the pyright error in pyright 1.1.385+
def apply_operation(item: T) -> T:
    return item.process("test")
```

**Expected behavior:** No type errors, since both `TypeA` and `TypeB` satisfy the `Processor` protocol and the function should preserve the input type.

**Actual behavior in v1.1.385+:**

```
Type "Processor" is not assignable to return type "T@apply_operation"
```

**Versions tested:**
- ✅ Works correctly: v1.1.384, v1.1.380, v1.1.378 (probably others)
- ❌ Broken: v1.1.385, v1.1.403 (probably others)

**Comment by erictraut** [comment]
Pyright's current behavior is correct here. There was a bug in older versions that I fixed. That explains why you saw a change in behavior.

For reference, pyright's current behavior also matches that of other type checkers like mypy.

The problem with the code is that the protocol doesn't guarantee that a value of the same class is returned by `process`. For example, the following class would satisfy this protocol:

```python
class TypeC:
    def process(self, arg: str) -> TypeB:
        return TypeB()
```

The use of `TypeC` would violate the return type of `apply_operation`.

**Comment by rijenkii** [comment]
@ibeckermayer you probably want to do this instead:
```diff
 class Processor(Protocol):
-    def process(self, arg: str) -> "Processor": ...
+    def process(self, arg: str) -> typing.Self: ...
```
[playground](https://pyright-play.net/?enableExperimentalFeatures=true&reportMissingModuleSource=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYBQVAxFAPIhJqoCGFU5AJgKY2kKbAM7CoAFUS8AggC4qURVD7AoCcKV6iAFMN4VgAGihsQaWVGEwQASigBaAHxQARJIQyX8pT6gheMACuICiW%2BsACQqISUgBC3koqahpawrrhxqbmltZ2Tq7uvLFeCr6K-kEhYQaRImJEZKm42g0k5BQ2CYpJ6o06egaZZhZWtg7OLg2aorheUAB0izRJbAgIFHAA%2BmAeIGwwSGAoANriFlNNIAC62kgwvBAW4nnOZ6XlAcGhdw-zvdNpFz3KwuGw0egAMSQAA9eNwoAADC4zEAImigSCwRCoDBYHB4KAAZXCtWiyOEuChsO4LWIZEonXeyl4qn%2BqXSgxMwxyY3yxIMFkW82WLJMaw2212%2B0OKE2wBhcNO5xSKKpcJuP0eEheEi6fk%2BVU1fxVgOBMFB4IkWhgwho0IAjFAALxi9ZbHa8PYHI5yhU0wrSbQ2MH%2BABuvA4m3gHm0DrBVGhACZna6JR6vTLfdTtIVYkGQ7xw5Ho7xY4mwUA)

**Comment by erictraut** [comment]
@rijenkii, thanks for pointing that out. Yes, that's a great solution.

**Comment by ibeckermayer** [comment]
Thanks for these helpful explanations @erictraut and @rijenkii. Just to lay it all out explicitly for the slow kids in the class (me):

This `Protocol` only dictates that the return type has to be another class which inherits from `Processor`, not necessarily the exact type of `self`:

```python
class Processor(Protocol):
    def process(self, arg: str) -> "Processor": ...
```

So pyright can't verify that the `T` passed in to `apply_operation` is the same `T` returned out:

```python
# TypeVar bound to the protocol
T = TypeVar("T", bound=Processor)

# Function that rightfully errors.
def apply_operation(item: T) -> T:
    return item.process("test")
```

@rijenkii's solution expresses what I was in fact trying to express, which is that the return value is the exact type of `self`:

```python
class Processor(Protocol):
    def process(self, arg: str) -> typing.Self: ...
```
